File: Java/EmployeeScheduler/src/com/company/scheduler/Scheduler.java
package com.company.scheduler;

import java.util.*;
public class Scheduler {
// List of days
    static final List<String> days =
            List.of("Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday");
// List of Duty in a Day
    static final List<String> shifts =
            List.of("MORNING","AFTERNOON","EVENING");
    static final int SHIFT_LIMIT = 2;
    Map<String, Map<String, List<Employee>>> weekPlan = new LinkedHashMap<>();
    Random random = new Random();
    public Scheduler() {
        for (String d : days) {
            Map<String, List<Employee>> map = new LinkedHashMap<>();
            for (String s : shifts) map.put(s, new ArrayList<>());
            weekPlan.put(d, map);
        }
    }
    public void createSchedule(List<Employee> staff) {

        // preference for 1st round
        for (int i = 0; i < days.size(); i++) {
            String day = days.get(i);

            for (Employee e : staff) {
                if (!e.canStillWork() || e.alreadyWorked(day)) continue;

                boolean placed = placeByPref(e, day);

                if (!placed) placed = placeAnywhere(e, day);

                if (!placed && i + 1 < days.size()) {
                    String nextDay = days.get(i + 1);
                    if (e.canStillWork() && !e.alreadyWorked(nextDay)) {
                        placeAnywhere(e, nextDay);
                    }
                }
            }
        }

        //  2  shift per day
        for (String d : days) {
            for (String s : shifts) {
                while (weekPlan.get(d).get(s).size() < 2) {
                    Employee pick = randomAvailable(staff, d, s);
                    if (pick == null) break;
                    assign(pick, d, s);
                }
            }
        }
    }
    private boolean placeByPref(Employee e, String day) {
        for (String s : e.prefMap.get(day)) {
            if (canAssign(e, day, s)) {
                assign(e, day, s);
                return true;
            }
        }
        return false;
    }
    private boolean placeAnywhere(Employee e, String day) {
        for (String s : shifts) {
            if (canAssign(e, day, s)) {
                assign(e, day, s);
                return true;
            }
        }
        return false;
    }
    private boolean canAssign(Employee e, String day, String shift) {
        if (!e.canStillWork()) return false;
        if (e.alreadyWorked(day)) return false;
        if (weekPlan.get(day).get(shift).size() >= SHIFT_LIMIT) return false;
        return true;
    }
    private void assign(Employee e, String day, String shift) {
        weekPlan.get(day).get(shift).add(e);
        e.daysWorked.add(day);
    }
    private Employee randomAvailable(List<Employee> staff, String day, String shift) {
        List<Employee> ok = new ArrayList<>();
        for (Employee e : staff) {
            if (canAssign(e, day, shift)) ok.add(e);
        }
        if (ok.isEmpty()) return null;
        return ok.get(random.nextInt(ok.size()));
    }
    public void printCompanySchedule() {
        System.out.println("\nFollowing is the weekly schedule for this week:\n");

        for (String d : days) {
            for (String s : shifts) {
                System.out.print(d + " " + pretty(s) + " - ");

                List<Employee> list = weekPlan.get(d).get(s);
                if (list.isEmpty()) {
                    System.out.println("-");
                } else {
                    for (int i = 0; i < list.size(); i++) {
                        System.out.print(list.get(i).name);
                        if (i < list.size() - 1) System.out.print(", ");
                    }
                    System.out.println();
                }
            }
            System.out.println();
        }
    }

    private String pretty(String s) {
        if (s.equals("MORNING")) return "Morning";
        if (s.equals("AFTERNOON")) return "Afternoon";
        return "Evening";
    }
}

File: Python/employee_scheduler_python/scheduler.py
# Umesh Dhakal
#1/28/2026
#MSCS632A4
import random

dayList = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
shiftList = ["MORNING", "AFTERNOON", "EVENING"]

MAX_PER_SHIFT = 2


class Employee:
    def __init__(self, empName):
        self.empName = empName
        self.dayPref = {}        # day -> [p1, p2, p3]
        self.workedDay = set()   # max 5 days


def createEmptySchedule():
    schedule = {}
    for day in dayList:
        schedule[day] = {}
        for shift in shiftList:
            schedule[day][shift] = []
    return schedule


def parseEmployeeNames(line):
    names = []
    parts = line.split(",")
    for p in parts:
        name = p.strip()
        if name:
            names.append(name)
    return names


def readThreeShift(msg):
    #user types 3 shifts in a day
    while True:
        line = input(msg).strip().upper().replace(",", " ")
        parts = line.split()

        if len(parts) != 3: 
            print("Enter 3 shifts like: MORNING AFTERNOON EVENING")
            continue

        if parts[0] in shiftList and parts[1] in shiftList and parts[2] in shiftList:
            return parts

        print("Only use MORNING, AFTERNOON, EVENING")


def canAssign(schedule, emp, day, shift):
    if len(emp.workedDay) >= 5:
        return False
    if day in emp.workedDay:
        return False
    if len(schedule[day][shift]) >= MAX_PER_SHIFT:
        return False
    return True


def assign(schedule, emp, day, shift):
    schedule[day][shift].append(emp.empName)
    emp.workedDay.add(day)


def buildSchedule(empList):
    schedule = createEmptySchedule()

    #preference first, then fix conflict
    for i in range(len(dayList)):
        day = dayList[i]

        for emp in empList:
            if day in emp.workedDay:
                continue
            if len(emp.workedDay) >= 5:
                continue

            done = False

            #try their 3 preferences
            for shift in emp.dayPref[day]:
                if canAssign(schedule, emp, day, shift):
                    assign(schedule, emp, day, shift)
                    done = True
                    break

            #try any shift same day
            if not done:
                for shift in shiftList:
                    if canAssign(schedule, emp, day, shift):
                        assign(schedule, emp, day, shift)
                        done = True
                        break

            if not done and i + 1 < len(dayList):
                nextDay = dayList[i + 1]
                for shift in shiftList:
                    if canAssign(schedule, emp, nextDay, shift):
                        assign(schedule, emp, nextDay, shift)
                        break

    #making minimum 2 employees per shift per day
    for day in dayList:
        for shift in shiftList:
            while len(schedule[day][shift]) < 2:
                available = [e for e in empList if canAssign(schedule, e, day, shift)]
                if not available:
                    break
                pick = random.choice(available)
                assign(schedule, pick, day, shift)

    return schedule


def printWeeklySchedule(schedule, empList):
    for emp in empList:
        print(f"\nWeekly Schedule for {emp.empName.upper()} is:")
        for day in dayList:
            empShift = "OFF"
            for shift in shiftList:
                if emp.empName in schedule[day][shift]:
                    empShift = shift
                    break
            print(f"{day} - {empShift}")


def printCompanySchedule(schedule):
    print("\nThe company schedule is:\n")

    for day in dayList:
        print(day)

        for shift in shiftList:
            names = schedule[day][shift]
            if names:
                print(f"{shift} : {', '.join(names)}")
            else:
                print(f"{shift} -")

        print()


def main():
    empCount = int(input("How many employee are there in a company - ").strip())
    
#Firstname1 Lastname1, Firstname2 Lastname2
    nameLine = input("Enter all the employee : ").strip()
    nameList = parseEmployeeNames(nameLine)

    while len(nameList) < empCount:
        extra = input("Add employee name : ").strip()
        if extra:
            nameList.append(extra)

    nameList = nameList[:empCount]

    empList = []
    for name in nameList:
        empList.append(Employee(name))

    for emp in empList:
        for day in dayList:
            msg = f"What is the preference schedule for {emp.empName} {day}: "
            emp.dayPref[day] = readThreeShift(msg)

    finalSchedule = buildSchedule(empList)

    printWeeklySchedule(finalSchedule, empList)
    printCompanySchedule(finalSchedule)


if __name__ == "__main__":
    main()
